### ISSUE TITLE
how to write and compile third module of rsyslog?

### ISSUE THREAD

**Issue by hxwang1024** [rsyslog/rsyslog#3533]
how to write third module of rsyslog and compile?

can you share a development tutorial, and how to compile third module

**Comment by hxwang1024** [comment]
There are a development tutorial document for writing and compiling a third module?

I wan to write a new third module for my need. But I don't know how to compile it into rsyslog.

**Comment by rgerhards** [comment]
create a PR, we can than give suggestions

**Comment by hxwang1024** [comment]
what PR is? 

rsyslog don't support to simplly write a third module for user defined need?

**Comment by rgerhards** [comment]
http://oss-watch.ac.uk/resources/pullrequest

**Comment by hxwang1024** [comment]
> create a PR, we can than give suggestions

I want to write a new third module on rsyslog, for eaxmple, I can write many third module on nginx by some development tutorial.
There are any development tutorial that how to write a now module on rsyslog, and how to compile a third module into rsyslog.

**Comment by rgerhards** [comment]
* https://www.slideshare.net/rainergerhards1/external-plugins
* details: https://github.com/rsyslog/rsyslog/blob/master/plugins/external/INTERFACE.md
* samples: https://github.com/rsyslog/rsyslog/blob/master/plugins/external/

**Comment by hxwang1024** [comment]
> * https://www.slideshare.net/rainergerhards1/external-plugins
> * details: https://github.com/rsyslog/rsyslog/blob/master/plugins/external/INTERFACE.md
> * samples: https://github.com/rsyslog/rsyslog/blob/master/plugins/external/

you  possibly not understand what I say.


I want to write a  third internal C language module in rsyslog,  and compile it into rsyslog by makefile.am/in  file. after, rsyslog can load and run it.  
for eaxmple, mmrm1stspace / imkafka / omstdout module.


now, I write a third C language module, But I don't know how to compile it into rsyslog.

There are a development tutorial document for writing and compiling a third module into rsyslog?

**Comment by rgerhards** [comment]
> There are a development tutorial document for writing and compiling a third module into rsyslog?

no, at least not that I am aware of - we usually suggest to use a simple module as template and change configure.ac/Makefile.am etc so that it fits.

**Comment by davidelang** [comment]
take a look at the modules in the contrib directory, you can write a module and 
add it to your local copy of the source and compile it yourself

If you want to have it upstream in rsyslog, you create a PR (see github for 
details) and the community helps review it and suggest improvements.

there are multiple types of modules in rsyslog, see the development guides that 
Rainer pointed you at for details.

I think you need to be a bit more clear about what you are trying to achieve.

David Lang

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant for the original question, but I am adding the answer here for people who arrive via search while we work through the older backlog. We do not close old issues lightly or just because of age.

The answer from the earlier thread still reflects the support boundary. For an external plugin, see the external plugin interface and examples under `plugins/external/`. For an in-tree/internal C module, there is not a separate standalone tutorial comparable to some other projects. The normal development approach is to copy a small existing module of the same kind, add the new module directory to the Automake/configure plumbing (`Makefile.am`, `configure.ac`, etc.), build it as part of the rsyslog source tree, and open a PR if it should be reviewed/upstreamed.

The `contrib/` directory is often the best place to study small optional modules. For example, `contrib/mmtaghostname` shows the basic shape of a loadable module with its own `Makefile.am`, configure option, documentation, and tests.

I am closing this as answered. If there is a specific current module skeleton or build-system error while adding a module, please open a new issue or PR with the attempted tree changes and build output.
